### PR TITLE
Add internal flag to disable property upgrades

### PR DESCRIPTION
**Pull request by asodja**
We can now pass "org.gradle.internal.disable.property.upgrades=true" to as system property to disable property upgrades.

We now add interface `BytecodeUpgradeInterceptor` to interceptors that we need to filter when a flag is set. 
We can probably do things a bit nicer in https://github.com/gradle/gradle/issues/26396, but it needs a bit more extra work.

Part of https://github.com/gradle/gradle-private/issues/4001

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGeneratorTest.groovy [hunk 1]
```diff
@@ -219,4 +220,45 @@ class InterceptJvmCallsGeneratorTest extends InstrumentationCodeGenTest {
...
+            }
+        """
+        def propertyUpgrade = source """
+            package org.gradle.test;
+
+            import org.gradle.api.provider.Property;
+            import org.gradle.internal.instrumentation.api.annotations.VisitForInstrumentation;
+            import org.gradle.internal.instrumentation.api.annotations.UpgradedProperty;
+
+            public abstract class Task {
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGeneratorTest.groovy:251]: :x: Let's not use `Task` as a type name here as it's easy to confuse with our API type. `CustomTask` should be a good alternative.

**Review by lptr — COMMENTED** [review]
This PR seems to be more about adding a flag to disable property upgrades. I think we should either split the PR, or describe it better, including both the title and the description of the PR.

I'm also not fond of using an environment variable to trigger the changed behavior. We usually use system properties under `org.gradle.unsafe` to enable/disable experimental features, and I don't see why we shouldn't be able to do it here as well.

**Review by lptr — COMMENTED** [review]
This PR seems to be more about adding a flag to disable property upgrades. I think we should either split the PR, or describe it better, including both the title and the description of the PR.

I'm also not fond of using an environment variable to trigger the changed behavior. We usually use system properties under `org.gradle.unsafe` to enable/disable experimental features, and I don't see why we shouldn't be able to do it here as well.

**Comment by lptr** [comment]
Let's close this as with recent changes we'd do this another way.

**Pull request closed without merging**